Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/tangential_relaxation.h

```diff
@@ -0,0 +1,326 @@
+      NamedParameters,
+      Static_boolean_property_map<edge_descriptor, false> // default (no constraint)
+    > ::type ECM;
+  ECM ecm = choose_parameter(get_parameter(np, internal_np::edge_is_constrained),
+                             Default_ECM());
+
+  typedef typename internal_np::Lookup_named_param_def <
+      internal_np::vertex_is_constrained_t,
+      NamedParameters,
+      Static_boolean_property_map<vertex_descriptor, false> // default (no constraint)
+    > ::type VCM;
+  VCM vcm = choose_parameter(get_parameter(np, internal_np::vertex_is_constrained),
+                             Static_boolean_property_map<vertex_descriptor, false>());
+
+  const bool relax_constraints = choose_parameter(get_parameter(np, internal_np::relax_constraints), false);
+  const unsigned int nb_iterations = choose_parameter(get_parameter(np, internal_np::number_of_iterations), 1);
+
+  typedef typename GT::Vector_3 Vector_3;
+  typedef typename GT::Point_3 Point_3;
+
+  auto check_normals = [&](vertex_descriptor v)
+  {
+    bool first_run = true;
+    Vector_3 prev = NULL_VECTOR, first = NULL_VECTOR;
+    halfedge_descriptor first_h = boost::graph_traits<TriangleMesh>::null_halfedge();
+    for (halfedge_descriptor hd : CGAL::halfedges_around_target(v, tm))
+    {
+      if (is_border(hd, tm)) continue;
+
+      Vector_3 n = compute_face_normal(face(hd, tm), tm, np);
+      if (n == CGAL::NULL_VECTOR) //for degenerate faces
+        continue;
+
+      if (first_run)
+      {
+        first_run = false;
+        first = n;
+        first_h = hd;
+      }
+      else
+      {
+        if (!get(ecm, edge(hd, tm)))
+          if (to_double(n * prev) <= 0)
+            return false;
+      }
+      prev = n;
+    }
+    if (!get(ecm, edge(first_h, tm)))
+      if (to_double(first * prev) <= 0)
+        return false;
+
+    return true;
+  };
+
+  typedef typename internal_np::Lookup_named_param_def <
+      internal_np::allow_move_functor_t,
+      NamedParameters,
+      internal::Allow_all_moves// default
+    > ::type Shall_move;
+  Shall_move shall_move = choose_parameter(get_parameter(np, internal_np::allow_move_functor),
+                                           internal::Allow_all_moves());
+
+  for (unsigned int nit = 0; nit < nb_iterations; ++nit)
+  {
+#ifdef CGAL_PMP_TANGENTIAL_RELAXATION_VERBOSE
+    std::cout << "\r\t(Tangential relaxation iteration " << (nit + 1) << " / ";
+    std::cout << nb_iterations << ") ";
+    std::cout.flush();
+#endif
+
+    typedef std::tuple<vertex_descriptor, Vector_3, Point_3> VNP;
+    std::vector< VNP > barycenters;
+
+    // at each vertex, compute vertex normal
+    std::unordered_map<vertex_descriptor, Vector_3> vnormals;
+    compute_vertex_normals(tm, boost::make_assoc_property_map(vnormals), np);
+
+    // at each vertex, compute barycenter of neighbors
+    for(vertex_descriptor v : vertices)
+    {
+      if (get(vcm, v) || halfedge(v, tm)==boost::graph_traits<TriangleMesh>::null_halfedge())
```

Comment:
Is the second check a check for "is_border(v, tm)"? I don't think it is a requirement for border vertices to have the canonical incident halfedge be a border (or not) halfedge.

Reply:
isolated vertex

Reply:
`is_isolated(v, tm)` for clarity?